include/storage/shared_monitor.hpp

```diff
@@ -0,0 +1,175 @@
+            if (!shmem.get_size(size) || size != internal_size + sizeof(Data))
+            {
+                auto message =
+                    boost::format("Wrong shared memory block '%1%' size %2%, expected %3% bytes") %
+                    (const char *)Data::name % size % (internal_size + sizeof(Data));
+                throw util::exception(message.str() + SOURCE_REF);
+            }
+
+            region = bi::mapped_region(shmem, bi::read_write);
+        }
+        catch (const bi::interprocess_exception &exception)
+        {
+            auto message = boost::format("No shared memory block '%1%' found, have you forgotten "
+                                         "to run osrm-datastore?") %
+                           (const char *)Data::name;
+            throw util::exception(message.str() + SOURCE_REF);
+        }
+    }
+
+    Data &data() const
+    {
+        auto region_pointer = reinterpret_cast<char *>(region.get_address());
+        return *reinterpret_cast<Data *>(region_pointer + internal_size);
+    }
+
+    mutex_type &get_mutex() const { return internal().mutex; }
+
+#if USE_BOOST_INTERPROCESS_CONDITION
+    template <typename Lock> void wait(Lock &lock) { internal().condition.wait(lock); }
+
+    void notify_all() { internal().condition.notify_all(); }
+#else
+    template <typename Lock> void wait(Lock &lock)
+    {
+        if (((internal().head + 1) & (buffer_size - 1)) == (internal().tail & (buffer_size - 1)))
+            throw util::exception(std::string("ring buffer is too small") + SOURCE_REF);
+
+        auto index = (internal().head++) & (buffer_size - 1);
+        auto semaphore = new (internal().buffer + index * sizeof(bi::interprocess_semaphore))
+            bi::interprocess_semaphore(0);
+        {
+            InvertedLock<Lock> inverted_lock(lock);
+            semaphore->wait();
+        }
+        semaphore->~interprocess_semaphore();
+    }
+
+    void notify_all()
+    {
+        bi::scoped_lock<mutex_type> lock(internal().mutex);
+        while (internal().tail != internal().head)
+        {
+            auto index = (internal().tail++) & (buffer_size - 1);
+            auto semaphore = reinterpret_cast<bi::interprocess_semaphore *>(
+                internal().buffer + index * sizeof(bi::interprocess_semaphore));
+            semaphore->post();
+        }
+    }
+#endif
+
+    static void remove() { bi::shared_memory_object::remove(Data::name); }
+
+  private:
+#if USE_BOOST_INTERPROCESS_CONDITION
+
+    static constexpr int internal_size = 128;
+
+    struct InternalData
+    {
+        mutex_type mutex;
+        bi::interprocess_condition condition;
+    };
+
+#else
+
+    static constexpr int buffer_size = 256;
+    static constexpr int internal_size = 4 * 4096;
+
+    struct InternalData
+    {
+        InternalData() : head(0), tail(0){};
+
+        std::size_t head, tail;
+        mutex_type mutex;
+        char buffer[buffer_size * sizeof(bi::interprocess_semaphore)];
```

Comment:
This could warrant a little bit of an explanation of what we are going with these semaphore: Implementing a conditional variable using a queue of semaphores.